helpers/getLocalizeResources.js

```diff
@@ -0,0 +1,355 @@
+
+		const previousVersion = cacheValue.headers.get(ETagHeader);
+		if (previousVersion !== currentVersion) {
+
+			console.log(`[Oslo] cache stale: ${resource}`);
+			fetchWithQueuing(resource).then(url => URL.revokeObjectURL(url));
+		}
+	}
+
+	const blob = await cacheValue.blob();
+	const objectUrl = URL.createObjectURL(blob);
+	return objectUrl;
+}
+
+function fetchWithPooling(resource) {
+
+	// At most one request per resource.
+
+	let promise = blobs.get(resource);
+	if (promise === undefined) {
+		promise = fetchWithCaching(resource);
+		blobs.set(resource, promise);
+	}
+	return promise;
+}
+
+function shouldUseBatchFetch() {
+
+	if (documentLocaleSettings === undefined) {
+		documentLocaleSettings = getDocumentLocaleSettings();
+	}
+
+	if (!documentLocaleSettings.oslo) {
+		return false;
+	}
+
+	// Only batch if we can do client-side caching, otherwise it's worse on each
+	// subsequent page navigation.
+
+	return Boolean(documentLocaleSettings.oslo.batch) && 'CacheStorage' in window;
+}
+
+function shouldUseCollectionFetch() {
+
+	if (documentLocaleSettings === undefined) {
+		documentLocaleSettings = getDocumentLocaleSettings();
+	}
+
+	if (!documentLocaleSettings.oslo) {
+		return false;
+	}
+
+	return Boolean(documentLocaleSettings.oslo.collection);
+}
+
+function setVersion(version) {
+
+	if (documentLocaleSettings === undefined) {
+		documentLocaleSettings = getDocumentLocaleSettings();
+	}
+
+	if (!documentLocaleSettings.oslo) {
+		return;
+	}
+
+	documentLocaleSettings.oslo.version = version;
+}
+
+function getVersion() {
+
+	if (documentLocaleSettings === undefined) {
+		documentLocaleSettings = getDocumentLocaleSettings();
+	}
+
+	const shouldReturnVersion =
+		documentLocaleSettings.oslo &&
+		documentLocaleSettings.oslo.version;
+	if (!shouldReturnVersion) {
+		return null;
+	}
+
+	return documentLocaleSettings.oslo.version;
+}
+
+function shouldFetchOverrides() {
+
+	const isOsloAvailable =
+		shouldUseBatchFetch() ||
+		shouldUseCollectionFetch();
+
+	return isOsloAvailable;
+}
+
+function fetchOverride(formatFunc, fetchFunc) {
+
+	let url, res;
+
+	if (shouldUseBatchFetch()) {
+
+		// If batching is available, pool requests together.
+
+		url = formatFunc(OverrideLanguage);
+		url = new URL(url).pathname;
+
+		res = fetchWithPooling(url);
+
+	} else /* shouldUseCollectionFetch() == true */ {
+
+		// Otherwise, fetch it directly and let the LMS manage the cache.
+
+		url = formatFunc(OverrideLanguage);
+		url = new URL(url).pathname;
+		url = documentLocaleSettings.oslo.collection + url;
+
+		res = Promise.resolve(url);
+
+	}
+
+	res = res.then(fetchFunc);
+	res = res.catch(coalesceToNull);
+
+	return res;
+}
+
+function coalesceToNull() {
+
+	return null;
+}
+
+export function __clearWindowCache() {
+
+	// Used to reset state for tests.
+
+	blobs.clear();
+	cache = undefined;
+	cachePromise = undefined;
+}
+
+export async function getLocalizeResources(
+	possibleLanguages,
+	filterFunc,
+	formatFunc,
+	fetchFunc
+) {
```

Comment:
How do you foresee this helper being used by a component -- could you give some sample code of what it might look like?

Reply:
Sure: https://github.com/BrightspaceHypermediaComponents/activities/compare/ppaskaris/oslo-get-overrides